Add exclusionFilters to USBDeviceRequestOptions

beaufortfrancois opened the pull request:
The new "exclusionFilters" option in `navigator.usb.requestDevice()` allows web developers to exclude some devices from the browser picker. It can be used to exclude devices that match a broader filter but are unsupported, or are already connected for instance.

```js
// Request access to a device from vendor ID 0xABCD. 
// The device with product ID 0x1234 has been reported as unsupported.
const device = await navigator.usb.requestDevice({
  filters: [{ vendorId: 0xABCD }],
  exclusionFilters: [{ vendorId: 0xABCD, productId: 0x1234 }],
});
```

Please have a look @reillyeon  

FIX: https://github.com/WICG/webusb/issues/232


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/beaufortfrancois/webusb/pull/233.html" title="Last updated on Jun 19, 2023, 9:49 AM UTC (899dc84)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/webusb/233/ed23354...beaufortfrancois:899dc84.html" title="Last updated on Jun 19, 2023, 9:49 AM UTC (899dc84)">Diff</a>

yume-chan commented:
Why do `requestDevice` and Peromission API treat the `excludionFilters` options differently? `requestDevice` doesn't allow it to be an empty array, and didn't check whether it was set in step 7.

reillyeon commented on index.bs:
```diff
@@ -664,28 +666,39 @@ steps <a>in parallel</a>:
+  6. Remove devices from |enumerationResult| if they do not <a>match a device
      filter</a> in <code>|options|.{{USBPermissionDescriptor/filters}}</code>.
-  6. Display a prompt to the user requesting they select a device from
+  7. Remove devices from |enumerationResult| if they do <a>match a device filter</a>
```
```suggestion
  7. Remove devices from |enumerationResult| if they <a>match a device filter</a>
```

reillyeon commented on index.bs:
```diff
@@ -2370,10 +2389,14 @@ defined as follows:
               to the next |device|.
-           2. Get the {{USBDevice}} representing |device| and add it to
+           2. If <code>|desc|.{{USBPermissionDescriptor/exclusionFilters}}</code>
+              is set and |device| does <a>match a device filter</a> in
```
```suggestion
              is set and |device| <a>matches a device filter</a> in
```

You might need to add an attribute to the definition of "match a device filter" for it to understand this version is the same definition. I forget how smart it is about this.

reillyeon reviewed (approved):
LGTM with nits. I agree that we probably should unify the behavior between `requestDevice()` and the Permissions API integration. Of course the latter is a bit aspirational since it isn't implemented in any browser engine.

beaufortfrancois commented:
> LGTM with nits. I agree that we probably should unify the behavior between `requestDevice()` and the Permissions API integration. Of course the latter is a bit aspirational since it isn't implemented in any browser engine.

Thank you! I've addressed your nits.

Would it be okay to land this as is and file a spec issue to unify the behaviour between `requestDevice()` and the Permissions API?

beaufortfrancois commented:
FYI I've started implementing this change in Chromium at https://chromium-review.googlesource.com/c/chromium/src/+/4614682.

Once the spec is updated, I'll start an intent to ship if that looks good to you @reillyeon

reillyeon commented:
> Would it be okay to land this as is and file a spec issue to unify the behaviour between `requestDevice()` and the Permissions API?

Yes.

beaufortfrancois commented:
@reillyeon Can you merge?
